### PR TITLE
Make controller sleep when not doing anything else

### LINKED ISSUE #49: On controller reset that is not power-on, issue a bus reset so modules reregister

### PR DESCRIPTION
**Pull request by timburke**
Controller now sleeps when the following two conditions are met:
- no tasks in the taskloop
- no rpc master call currently in progress

The flash memory is disabled in sleep to save power.  It is automatically reenabled upon first use by the memory module and there is a wait-state incorporated before the first write to the flash since it needs at least 10 ms to warm up and allow writing.  All flash memory writes now wait until they complete before returning so that we cannot sleep (and disable the memory) while the flash memory was programming itself.

Also made the controller reset the MIB bus whenever it resets so that all modules reregister themselves. This means the controller will always have a correct picture of which modules are on the bus.

Closes #49, #22

### CODE REVIEW

**Comment by amcgee** [comment]
I'm going to test this out a bit before merging, generally looks good though.

File: momo_modules/mainboard/src/core/mainboard_reset_handler.c [hunk 1]
```diff
@@ -20,8 +20,10 @@ void handle_all_resets_before(unsigned int type)
 {
     //Add code here that should be called before all other reset code
     disable_unneeded_peripherals();
+    init_memory();
+    enable_memory(1);
     configure_interrupts();
-    oscillator_init();
+    //oscillator_init();
```
Comment (amcgee) [inline momo_modules/mainboard/src/core/mainboard_reset_handler.c:26]: Should this be removed completely?

File: momo_modules/shared/pic24/src/core/task_manager.c [hunk 2]
```diff
@@ -61,12 +80,14 @@ void taskloop_loop()
 {
     while(1)
     {
-        while( taskloop_process_one() )
+        while (taskloop_process_one())
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/core/task_manager.c:83]: We should decide on a standard coding style and enforce it.  the `if ( condition )` style has been hard-wired into my system so I find it more readable, but I think `if (condition)` is more standard.  Even better, maybe we should automate the enforcement of whichever standard by using something like [ClangFormat](http://clang.llvm.org/docs/ClangFormat.html).  If we go that route we could also possibly consider using the [clang/llvm static analyzer](http://clang-analyzer.llvm.org/), though I don't know how well it would work with microchip's version of C.

File: momo_modules/shared/pic24/src/core/task_manager.h [hunk 3]
```diff
@@ -15,11 +15,21 @@
 #define kMAXTASKS 16 //NB Must be a power of 2 since it will be used for a ringbuffer
 
 typedef void (*task_callback)(void);
+typedef int (*sleep_callback)(int);
 
 enum
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/core/task_manager.h]: if we typedef this enum I think we can reference it by name as an argument type, it's more explicit and self-documenting that way.

File: momo_modules/shared/pic24/src/core/task_manager.h [hunk 4]
```diff
@@ -30,8 +40,9 @@ typedef struct
 
 typedef struct
 {
-    task_item taskdata[kMAXTASKS];
-    ringbuffer tasks;
+    task_item 		taskdata[kMAXTASKS];
+    ringbuffer 		tasks;
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/core/task_manager.h]: Nitpick: alignment

File: momo_modules/shared/pic24/src/modules/memory.c [hunk 5]
```diff
@@ -51,6 +69,56 @@ void configure_SPI() {
   DISABLE_MEMORY(); //idle state of SS is high
 }
 
+void enable_memory(uint8 for_writing)
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/modules/memory.c]: This name is now overloaded with the all-caps macro `ENABLE_MEMORY`, should we rename the macro?

Also, can for_writing be semantically either a bool or an enum type if we expect other options at some point?

File: momo_modules/shared/pic24/src/modules/memory.c [hunk 6]
```diff
@@ -149,12 +221,16 @@ void mem_write_aligned(const uint32 addr, const BYTE *data, unsigned int length)
     spi_transfer(data[i]);
 
   DISABLE_MEMORY();
+
+  mem_wait_while_writing();
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/modules/memory.c:228]: So technically this changes the behavior of the `mem_write()` function, previously it would asynchronously write and only block if there was another write currently in progress, was this change intentional?

File: tools/pymomo/commander/proxy/controller.py [hunk 7]
```diff
@@ -100,6 +100,13 @@ def get_debug_value(self):
 		flag = res['ints'][0]
 		return flag
 
+	def set_sleep(self, val):
+		if val:
+			arg = 1
+		else:
+			arg = 0
```
Comment (amcgee) [inline tools/pymomo/commander/proxy/controller.py]: This block could technically be written as just `int( bool( val ) )` for simplicity, or even just `res = self.rpc(42, 0x0E, bool( val ) )` and rpc should do the right thing by implicitly converting it to a 0 or 1.

File: momo_modules/shared/pic24/src/modules/memory.h [hunk 8]
```diff
@@ -17,10 +20,16 @@
...
+typedef struct
+{
+	unsigned int write_wait : 1;
+	unsigned int enabled : 1;
+	unsigned int reserved : 14;
+} memory_config;
 
 void mem_write(uint32 addr, const BYTE* data, unsigned int length );
 void mem_write_aligned(const uint32 addr, const BYTE* data, unsigned int length);
+void mem_wait_while_writing();
```
Comment (amcgee) [inline momo_modules/shared/pic24/src/modules/memory.h]: It would be nice to not have internal static functions in the header, did it need to come out for tests to work?

File: momo_modules/shared/pic24/src/core/task_manager.c (hunk 2 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/core/task_manager.c:83]: I agree with this comment whole-heartedly.  We should set up a git commit hook or something to format the code upon commit.  I've been particularly bad in this regard trying to get things done quickly.  I like if () and brackets on the line following the control structure but that's just me.  We should start a coding standard document

Comment (amcgee) [inline momo_modules/shared/pic24/src/core/task_manager.c:83]: Agreed.  Do you like spaces after the parentheses or not?

On Sat, Apr 12, 2014 at 2:21 PM, timburke notifications@github.com wrote:

> In momo_modules/shared/pic24/src/core/task_manager.c:
> 
> > @@ -61,12 +80,14 @@ void taskloop_loop()
> >  {
> >      while(1)
> >      {
> > -        while( taskloop_process_one() )
> > -        while (taskloop_process_one())
> 
> I agree with this comment whole-heartedly. We should set up a git commit
> hook or something to format the code upon commit. I've been particularly
> bad in this regard trying to get things done quickly. I like if () and
> brackets on the line following the control structure but that's just me. We
> should start a coding standard document
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/WellDone/MoMo-Firmware/pull/60/files#r11562059
> .

## 

Austin McGee
austinwm@gmail.com
(510) 200-8486

File: momo_modules/shared/pic24/src/core/task_manager.h (hunk 4 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/core/task_manager.h]: Agreed, alignment was bad and I should feel bad.  The issue was a mixture of spaces and tabs that made it seem aligned right on my computer.  Should we decide on 4 spaces per tab and enforce no tabs?

Comment (amcgee) [inline momo_modules/shared/pic24/src/core/task_manager.h]: I prefer 2 spaces, but 4 is ok and yeah I think no tabs.  Another task for
a post-commit format hook.

On Sat, Apr 12, 2014 at 2:24 PM, timburke notifications@github.com wrote:

> In momo_modules/shared/pic24/src/core/task_manager.h:
> 
> > @@ -30,8 +40,9 @@ typedef struct
> > 
> >  typedef struct
> >  {
> > -    task_item taskdata[kMAXTASKS];
> > -    ringbuffer tasks;
> > -    task_item      taskdata[kMAXTASKS];
> > -    ringbuffer         tasks;
> 
> Agreed, alignment was bad and I should feel bad. The issue was a mixture
> of spaces and tabs that made it seem aligned right on my computer. Should
> we decide on 4 spaces per tab and enforce no tabs?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/WellDone/MoMo-Firmware/pull/60/files#r11562064
> .

## 

Austin McGee
austinwm@gmail.com
(510) 200-8486

File: momo_modules/shared/pic24/src/core/task_manager.c (hunk 2 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/core/task_manager.c:83]: I personally don't.

File: momo_modules/shared/pic24/src/core/task_manager.h (hunk 3 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/core/task_manager.h]: Agreed and fixed.

File: momo_modules/shared/pic24/src/modules/memory.c (hunk 5 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/modules/memory.c]: Agreed and fixed.  Updated memory functions to have common naming scheme like the other modules: mod_functioname().

File: momo_modules/shared/pic24/src/modules/memory.c (hunk 6 above)
Comment (timburke) [inline momo_modules/shared/pic24/src/modules/memory.c:228]: It was intentional so that we can better reason about the state of the memory module in taskloop_loop().  Without this, you would have to check if the module was still writing before removing power to it and going to sleep.  Also, the module locks itself while writing so you can't read from it either, meaning that all read functions would have to have a mem_wait_while_writing check in front of them too.  

It seemed that the clearer thing to do was simply have the write be synchronous and finish by the time the function returns.

Comment (amcgee) [inline momo_modules/shared/pic24/src/modules/memory.c:228]: I buy that, sounds good.

On Sat, Apr 12, 2014 at 2:46 PM, timburke notifications@github.com wrote:

> In momo_modules/shared/pic24/src/modules/memory.c:
> 
> > @@ -149,12 +221,16 @@ void mem_write_aligned(const uint32 addr, const BYTE *data, unsigned int length)
> >      spi_transfer(data[i]);
> > 
> >    DISABLE_MEMORY();
> > +
> > -  mem_wait_while_writing();
> 
> It was intentional so that we can better reason about the state of the
> memory module in taskloop_loop(). Without this, you would have to check if
> the module was still writing before removing power to it and going to
> sleep. Also, the module locks itself while writing so you can't read from
> it either, meaning that all read functions would have to have a
> mem_wait_while_writing check in front of them too.
> 
> It seemed that the clearer thing to do was simply have the write be
> synchronous and finish by the time the function returns.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/WellDone/MoMo-Firmware/pull/60/files#r11562142
> .

## 

Austin McGee
austinwm@gmail.com
(510) 200-8486

File: momo_modules/mainboard/src/core/mainboard_reset_handler.c (hunk 1 above)
Comment (timburke) [inline momo_modules/mainboard/src/core/mainboard_reset_handler.c:26]: It should as it currently doesn't do anything.  The secondary oscillator can be used for longterm timebase stability in the RTCC but is currently not.  I thought that it was b/c I enabled it in oscillator_init but it turns out that you need to set a config fuse to tell the RTCC to use the socs instead of its own RC oscillator.

**Comment by amcgee** [comment]
I've tested this and it seems to be working fine for me!
